addons/account/models/account_move.py

```diff
@@ -2474,9 +2474,11 @@ def _post(self, soft=True):
             (partner | partner.commercial_partner_id)._increase_rank('supplier_rank', count)
 
         # Trigger action for paid invoices in amount is zero
-        to_post.filtered(
+        invoices_zero = to_post.filtered(
             lambda m: m.is_invoice(include_receipts=True) and m.currency_id.is_zero(m.amount_total)
-        ).action_invoice_paid()
+        )
+        for company in invoices_zero.mapped('company_id'):
+            invoices_zero.filtered(lambda i: i.company_id == company).with_company(company).action_invoice_paid()
```

Comment:
The business code is not supposed to run using the current company of the business object. You can't depends of the current user company in any way when dealing with accounting. So IMO, the code is fine like this. If you are facing some inconsitency regarding the `ir.property`, this is the responsability of the others modules to add the `with_company` when accessing the property.

Reply:
> The business code is not supposed to run using the current company of the business object. 

Isn't that's *exactly* what business code is supposed to do? The reference company when processing a record and creating new related records is the company of the object, not the company of the user running the action.

This used to be done by setting `force_company` in the context, which would allow most company-dependent things to behave correctly, including property field access, etc. This has been replaced by `with_company()`, for the same purpose.

And yes, `env.company` is not the same thing as `env.user.company_id`.

/cc @qdp-odoo (since @smetl is afk ^^)

Reply:
There is absolutely nothing inside the orm batching the records by company. Therefore, you should never based your business code on the current company since the `self` in your `compute` methods / business methods could contains records from multiple-companies. Also, the orm is not doing any implicit `with_company` so to be consistent with the diff, we should apply the same logic everywhere because I don't see any reason to apply such change to only this specific method.